internal/controllers/utils/utils.go

```diff
@@ -14,36 +14,64 @@ import (
 	"github.com/kong/kubernetes-ingress-controller/v2/internal/annotations"
 )
 
-// HasAnnotation is a helper function to determine whether an object has a given annotation, and whether it's
-// to the value provided.
-func HasAnnotation(obj client.Object, key, expectedValue string) bool {
-	foundValue, ok := obj.GetAnnotations()[key]
-	return ok && foundValue == expectedValue
+const defaultIngressClassAnnotation = "ingressclass.kubernetes.io/is-default-class"
+
+// MatchesClass is a helper function to determine whether an object has a given ingress class or no class if the given
+// class is the default class
+func MatchesClass(obj client.Object, class string, isDefault bool) bool {
+	annotation := annotations.IngressClassKey
+	if _, ok := obj.(*knative.Ingress); ok {
+		annotation = annotations.KnativeIngressClassKey
+	}
+	foundValue, ok := obj.GetAnnotations()[annotation]
+	if !ok && isDefault {
+		return true
+	}
+	return ok && foundValue == class
+}
+
+// IsDefaultIngressClass returns whether an IngressClass is the default IngressClass
+func IsDefaultIngressClass(obj client.Object) bool {
+	if ingressClass, ok := obj.(*netv1.IngressClass); ok {
+		return ingressClass.ObjectMeta.Annotations[defaultIngressClassAnnotation] == "true"
+	}
+	return false
 }
 
 // MatchesIngressClassName indicates whether or not an object indicates that it's supported by the ingress class name provided.
-func MatchesIngressClassName(obj client.Object, ingressClassName string) bool {
+func MatchesIngressClassName(obj client.Object, ingressClassName string, isDefault bool) bool {
```

Comment:
This seems so closely related to `MatchesClass` that I'm not sure that we should have separate functions as that would seem to just cause confusion. It's possible I'm missing something in the minutia of this but my current suggestion is to combine these into a single function.

Reply:
It's basically an artifact of the original `HasAnnotation()` independently answering only the "does value match other value" question. There's not any particular reason for them to be separate, I just wrote an additional function over top rather than expanding the existing one.